src/Dialect/ONNX/ONNXOps/Math/Reduction.cpp

```diff
@@ -74,75 +74,87 @@ LogicalResult ONNXGenericReductionOpShapeHelper<OP_TYPE>::customComputeShape(
-  ONNXReduceSumOp reduceOp = llvm::cast<ONNXReduceSumOp>(op);
-  ONNXReduceSumOpAdaptor operandAdaptor(operands, op->getAttrDictionary());
-  DimsExpr axes;
-  if (isFromNone(operandAdaptor.axes())) {
-    // Default will be used.
-  } else if (getONNXConstantOp(operandAdaptor.axes())) {
-    createIE->getIntFromArrayAsSymbols(operandAdaptor.axes(), axes);
+  // Handle simple case where axes is an attribute.
+  if constexpr (!isAxesInput<OP_TYPE>) {
+    createIE->getIntFromArrayAsLiterals(operandAdaptor.axesAttr(), axes);
+    return customComputeShape(axes, /*noopWithEmptyAxes*/ false);
   } else {
-    // When the axis is dynamic, try to infer the rank of output tensor
-    int64_t dataRank = createIE->getShapedTypeRank(operandAdaptor.data());
-    int64_t axlesSize = createIE->getArraySize(operandAdaptor.axes());
-    if (!operandAdaptor.keepdims() && axlesSize < 0 /*undef shape*/) {
-      // Even though we did not compute the shape in ShapeHelper, return success
-      // as we gen code for dyn sizes too. Ideally, we would have the code here
-      // to compte the shape, it is currently residing in Krnl lowering.
-
-      // In fact, this case here requires further handling of unranked tensors,
-      // which I doubt we support. But there is currently a lit tests
-      // (onnx/onnx_shape_inference.mlir's test_reduce_sum_5) expect the *xf32
-      // sizes. So leave reporting success here, even though I really suspect we
-      // should report failure. return op->emitError("does not support unranked
-      // reduction output");
-      return success();
-    }
-    // With keep dim, by def the output has the same rank as the input; without
-    // keep dim, we remove 1 rank per value in the 1D axes tensor.
-    int64_t outputRank =
-        operandAdaptor.keepdims() ? dataRank : dataRank - axlesSize;
-    assert(outputRank >= 0 && "expected to keep at least one dim");
-
-    // Interesting idea below, namely to reuse info from output, or if not
-    // there, from input putting question mark in there. Not sure if successful,
-    // if it is, it should be generalized to all ops.
-
-    if (reduceOp.getResult().getType().isa<RankedTensorType>()) {
-      // Have already some shapes, keep them in ShapeHelper
-      DimsExpr outputDims;
-      createIE->getShapeAsDims(reduceOp.getResult(), outputDims);
+
+    // Handle case where reduction op has axes as input.
+    if (isFromNone(operandAdaptor.axes())) {
+      // Default will be used.
+    } else if (getONNXConstantOp(operandAdaptor.axes())) {
+      createIE->getIntFromArrayAsSymbols(operandAdaptor.axes(), axes);
+    } else {
+      // When the axis is dynamic, try to infer the rank of output tensor
+      int64_t dataRank = createIE->getShapedTypeRank(operandAdaptor.data());
+      int64_t axlesSize = createIE->getArraySize(operandAdaptor.axes());
+      if (!operandAdaptor.keepdims() && axlesSize < 0 /*undef shape*/) {
```

Comment:
Avoid using "<0" for dynamic dim.
Here we return success without improving the shape info, hoping further transformation could provide extra info. If that does not happen, an assertion error for unranked tensor should arise at the pre-lowering check. I tried the following example with main branch(without this PR):
```
func.func private @main_graph(%arg0: tensor<3x2x2xf32>, %arg1: tensor<?xi64>) -> tensor<*xf32> {
  %0 = "onnx.ReduceSum"(%arg0, %arg1) {keepdims = 0 : si64, noop_with_empty_axes = 1 : si64} : (tensor<3x2x2xf32>, tensor<?xi64>) -> tensor<*xf32>
  return %0 : tensor<*xf32>
}
```
I got the following result for shape inference:
```
func.func private @main_graph(%arg0: tensor<3x2x2xf32>, %arg1: tensor<?xi64>) -> tensor<f32> {
  %0 = "onnx.ReduceSum"(%arg0, %arg1) {keepdims = 0 : si64, noop_with_empty_axes = 1 : si64} : (tensor<3x2x2xf32>, tensor<?xi64>) -> tensor<f32>
  return %0 : tensor<f32>
}
```
That's weird. 
Of course, the keep dims == 0 caused assertion error in lowering.
```
loc("reducesum.mlir":3:8): error: not keepdims() not implemented
```
These issues need to be addressed in future PR.

Reply:
Thanks for the review @chentong319,

I actually didn't write that code. I just templated the existing implementation to extend its support to `ReduceProd`, `ReduceMean`, `ReduceMax`, `ReduceMin`.

I do agree with you points. However as you suggested I think addressing them in a future PR makes sense, as they are slightly unrelated with the templating I did. 

The `git diff` of this file doesn't do a good job of showing that only a few line were changed.